pkg/types/chain_reader.go

```diff
@@ -51,21 +75,260 @@ type BoundContract struct {
+	OR
+)
+
+func (op BooleanOperator) String() string {
+	switch op {
+	case AND:
+		return "AND"
+	case OR:
+		return "OR"
+	default:
+		return "Unknown"
+	}
+}
+
+// BooleanExpression allows nesting of boolean expressions with different BooleanOperator's.
+type BooleanExpression struct {
+	// should have minimum length of two
+	Expressions []Expression
+	BooleanOperator
+}
+
+func NewBooleanExpression(operator BooleanOperator, expressions ...Expression) Expression {
+	return Expression{
+		BooleanExpression: BooleanExpression{Expressions: expressions, BooleanOperator: operator},
+	}
+}
+
+// BlockFilter is a primitive of QueryFilter that filters search in comparison to block number.
+type BlockFilter struct {
+	Block    uint64
+	Operator ComparisonOperator
+}
+
+func NewBlockPrimitive(block uint64, operator ComparisonOperator) Expression {
+	return Expression{
+		Primitive: &BlockFilter{Block: block, Operator: operator},
+	}
+}
+
+func (f *BlockFilter) Accept(visitor Visitor) {
+	visitor.VisitBlockFilter(*f)
+}
+
+// AddressFilter is a primitive of QueryFilter that filters search to results that contain address in Addresses.
+type AddressFilter struct {
+	Addresses []string
+}
+
+func NewAddressesPrimitive(addresses ...string) Expression {
+	return Expression{
+		Primitive: &AddressFilter{Addresses: addresses},
+	}
+}
+
+func (f *AddressFilter) Accept(visitor Visitor) {
+	visitor.VisitAddressFilter(*f)
+}
+
+type Confirmations int32
+
+const (
+	Finalized   = Confirmations(0)
+	Unconfirmed = Confirmations(1)
+)
+
+// ConfirmationsFilter is a primitive of QueryFilter that filters search to results that have a certain level of confirmation.
+// Confirmations map to different concepts on different blockchains.
+type ConfirmationsFilter struct {
+	Confirmations
+}
+
+func NewConfirmationsPrimitive(confs Confirmations) Expression {
+	return Expression{
+		Primitive: &ConfirmationsFilter{Confirmations: confs},
+	}
+}
+
+func (f *ConfirmationsFilter) Accept(visitor Visitor) {
+	visitor.VisitConfirmationFilter(*f)
+}
+
+// TimestampFilter is a primitive of QueryFilter that filters search in comparison to timestamp.
+type TimestampFilter struct {
+	Timestamp uint64
+	Operator  ComparisonOperator
+}
+
+func NewTimestampPrimitive(timestamp uint64, operator ComparisonOperator) Expression {
+	return Expression{
+		Primitive: &TimestampFilter{timestamp, operator},
+	}
+}
+
+func (f *TimestampFilter) Accept(visitor Visitor) {
+	visitor.VisitTimestampFilter(*f)
+}
+
+// TxHashFilter is a primitive of QueryFilter that filters search to results that contain txHash.
+type TxHashFilter struct {
+	TxHash string
+}
+
+func NewTxHashPrimitive(txHash string) Expression {
+	return Expression{
+		Primitive: &TxHashFilter{txHash},
+	}
+}
+
+func (f *TxHashFilter) Accept(visitor Visitor) {
+	visitor.VisitTxHashFilter(*f)
+}
+
+// Where eg. usage:
+// queryFilter, err := Where(
+//
+//		NewTxHashPrimitive("0xHash"),
+//		NewBooleanExpression("OR",
```

Comment:
Some helpers for `OR` & `AND` would be nice:
```suggestion
//		NewOr(
```
And external callers will have to include the package, so it might be worth declaring these in a different place, e.g. `package expr`?
```suggestion
//		expr.NewOr(
```
I think dropping `New` would be fine in that context too:
```suggestion
//		expr.Or(
```

Or maybe `package query`? e.g. `query.Expression`, `query.Or(`, `query.Filter`